src/transformers/agents/tools.py

```diff
@@ -808,3 +844,37 @@ def __init__(self, collection_slug: str, token: Optional[str] = None):
         self._collection = get_collection(collection_slug, token=token)
         self._hub_repo_ids = {item.item_id for item in self._collection.items if item.item_type == "space"}
         self.tools = {Tool.from_hub(repo_id) for repo_id in self._hub_repo_ids}
+
+
+def tool(tool_function: Callable) -> Tool:
+    """
+    Decorator that turns a function into an instance of a specific Tool subclass
+
+    Args:
+        tool_function: Your function. Should have type hints for each input and a type hint for the output.
+        Should also have a docstring description including an 'Args:' part where each argument is described.
+    """
+    parameters = get_json_schema(tool_function)["function"]
+    if "return" not in parameters:
+        raise TypeHintParsingException("Tool return type not found: make sure your function has a return type hint!")
+    class_name = f"{parameters['name'].capitalize()}Tool"
+
+    class SpecificTool(Tool):
+        name = parameters["name"]
+        description = parameters["description"]
+        inputs = parameters["parameters"]["properties"]
+        output_type = parameters["return"]["type"]
+
+        @wraps(tool_function)
+        def forward(self, *args, **kwargs):
+            return tool_function(*args, **kwargs)
+
+    original_signature = inspect.signature(tool_function)
+    new_parameters = [inspect.Parameter("self", inspect.Parameter.POSITIONAL_OR_KEYWORD)] + list(
+        original_signature.parameters.values()
+    )
+    new_signature = original_signature.replace(parameters=new_parameters)
+    SpecificTool.forward.__signature__ = new_signature
+
+    SpecificTool.__name__ = class_name
+    return SpecificTool()
```

Comment:
We discussed this a little on Slack but I've been thinking about it: Do you think users will get confused because of the different approaches taken in chat templates / agents?

Chat templates: Tools are passed as **JSON schema**. Users can also directly pass Python functions, and these will be automatically converted to JSON schema.

Agents: Tools are passed as `Tool` objects. Users can convert functions to `Tool` objects using the `@tool` decorator.

Do we need both approaches? We could have a more unified API by, for example, removing the `tool` decorator and instead converting functions to `Tool` objects if they're passed to agents, the same way we do in chat templates. That way, even if the internal objects are different, the API for users would be the same.

Reply:
Thank you for taking a look @Rocketknight1 !

So it seems to me that a fundamental difference is that in Chat templates, you can pass a JSON scheme : this is completely impossible for our tools since the JSON cannot encode the internal logic of the tool's forward pass.

But we can add the functionality to accept raw functions as tools, and only convert them to tools upon agent initialization! So we would support both `Tool` objects and passing raw functions.

I think we can keep supporting the decorator as well, since applying a decorator directly to the tool has the advantage to let the users try out the tool before it's passed to the agent initialization (which I personnally often do to make sure they work properly). Anyway this decorator is just the `tool` function, which we will need in any case for converting functions to Tools.

So I'd say let's keep the usage of `tool` as a decorator and add support for raw functions on agent initialization (which will also use `tool` behind the hood for conversion to `Tool` objects)! Wdyt as well @LysandreJik ?

Reply:
Ok let's merge this for now, since the solution you propose @Rocketknight1 only needs to add support for raw functions in the `tools` list upon agent initialization, which we can do in a later PR!

Reply:
Sure!